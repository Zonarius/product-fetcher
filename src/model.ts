import { Observable } from "rxjs";

export interface Fetcher<T extends Record<string, any>> {
  matches(url: string): boolean;
  fetchProducts(url: string): Observable<T>
}
