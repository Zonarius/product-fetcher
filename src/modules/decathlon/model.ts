export interface DecathlonProduct {
  name: string;
  // In Euro
  price: number;
  // In kg
  weight: number;
  technicalData: Record<string, string>;
}