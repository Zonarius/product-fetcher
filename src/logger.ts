import log4js from 'log4js';

log4js.configure({
  appenders: {
    out: { type: 'stderr' }
  },
  categories: {
    default: { appenders: ["out"], level: "info" }
  }
});
export const logger = log4js.getLogger();
