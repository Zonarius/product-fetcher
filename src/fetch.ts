import fetch from 'node-fetch';
import { HTMLElement, parse } from 'node-html-parser';
import { defer, from, map, Observable } from 'rxjs';
import { logger } from './logger';

export function getText(url: string): Promise<string> {
  logger.info(`Getting ${url}`);
  return fetch(url).then(r => r.text());
}

export function getParsed(url: string): Promise<HTMLElement> {
  return getText(url).then(parse);
}

export function getParsedObs(url: string): Observable<HTMLElement> {
  return defer(() => from(getParsed(url)));
}