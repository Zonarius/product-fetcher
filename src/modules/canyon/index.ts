import { HTMLElement } from "node-html-parser";
import { from, mergeMap, Observable, of, switchMap } from "rxjs";
import { getParsedObs } from "../../fetch";
import { Fetcher } from "../../model";
import { parseNumber } from "../../parse";
import { createRecord } from "../../util";
import { CanyonProduct } from "./model";

export function matches(url: string) {
  return /^https?:\/\/www\.canyon\.com/.test(url);
}

export function fetchProducts(url: string): Observable<CanyonProduct> {
  return getParsedObs(url).pipe(
    switchMap(document => {
      if (document.querySelector("#section-product-grid")) {
        return from(getProductListUrls(document))
          .pipe(
            mergeMap(fetchProducts)
          )
      } else {
        return of({ ...parseProduct(document), url } as CanyonProduct);
      }
    })
  )
}

export function getProductListUrls(document: HTMLElement): string[] {
  return document.querySelectorAll("a.productTile__link")
    .map(el => el.getAttribute("href")!);
}

export function parseProduct(document: HTMLElement): Partial<CanyonProduct> {
  return {
    name: document.querySelector("h1.productDescription__productName").rawText.trim(),
    price: parseNumber(document.querySelector(".productDescription__priceSale").rawText)!,
    weight: getWeight(document),
  }
}

function getWeight(document: HTMLElement): number {
  return parseNumber(getCharacteristics(document)["Weight"])!;
}

function getCharacteristics(document: HTMLElement): Record<string, string> {
  return createRecord(
    document.querySelectorAll(".productDescription__characteristicsListItem")
      .map(item => {
        const [label, detail] = item.querySelectorAll("div");
        return [label.rawText.trim(), detail.rawText.trim()];
      })
  );
}

const moduletypeCheck = { matches, fetchProducts } as Fetcher<CanyonProduct>