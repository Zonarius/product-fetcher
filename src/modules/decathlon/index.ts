import { HTMLElement } from "node-html-parser";
import { from, mergeMap, Observable, of, switchMap } from "rxjs";
import { getParsedObs } from "../../fetch";
import { Fetcher } from "../../model";
import { parseNumber } from "../../parse";
import { createRecord } from "../../util";
import { DecathlonProduct } from "./model";

export function matches(url: string) {
  return /^https?:\/\/www.decathlon./.test(url);
}

export function fetchProducts(url: string): Observable<DecathlonProduct> {
  return getParsedObs(url).pipe(
    switchMap(document => {
      if (document.querySelector("#category-hits")) {
        return from(getProductListUrls(document))
          .pipe(
            mergeMap(fetchProducts)
          )
      } else {
        return of(parseProduct(document));
      }
    })
  )
}

export function getProductListUrls(document: HTMLElement): string[] {
  return document.querySelectorAll(".thumbnail--product a.thumbnail")
    .map(el => el.getAttribute("href")!);
}

export function parseProduct(document: HTMLElement): DecathlonProduct {
  const technicalData = parseTechnicals(document);
  return {
    name: document.querySelector("h1.js-title").rawText,
    price: parseNumber(document.querySelector(".price-text").rawText)!,
    weight: parseNumber(technicalData["Gewicht"])!,
    technicalData
  }
}

function parseTechnicals(document: HTMLElement): Record<string, string> {

  return createRecord(document.querySelectorAll("#product-technicals .row")
    .map(parseTechnicalsRow));
}

function parseTechnicalsRow(row: HTMLElement): [string, string] {
  const [rawTitle, rawText] = row.querySelectorAll("div")
  return [rawTitle.rawText.trim(), rawText.rawText.trim()];
}

const moduletypeCheck = { matches, fetchProducts } as Fetcher<DecathlonProduct>
