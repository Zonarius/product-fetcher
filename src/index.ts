#!/usr/bin/env -S node -r "ts-node/register"

import { Command, Option } from 'commander';
import { take, toArray } from 'rxjs';
import { logger } from './logger';
import modules from './modules';

const { version } = require("../package.json");
const program = new Command();

process.env["NODE_OPTIONS"] = "--insecure-http-parser";

program
  .name("product-fetcher")
  .description("Fetches product data from various sources.")
  .version(version)
  .option("-v, --verbose", "Verbose output")
  .option("--list-sources", "Lists all available sources")
  .option("-l, --limit <limit>", "The maximum number of products to fetch", Number, -1)
  .argument("[url]", "The url to fetch products from")
  .action(main)
  .parse();

async function main(url: string) {
  const options = program.opts();
  logger.level = options.verbose
    ? "debug"
    : "info";
    
  if (options.listSources) {
    Object.keys(modules).forEach(m => console.log(m));
    return;
  }

  if (!url) {
    program.help({ error: true })
  }

  const matched = Object.values(modules).find(f => f.matches(url));
  if (!matched) {
    logger.error("No module matches the URL!");
    return;
  }

  const obs = options.limit >= 0
    ? matched.fetchProducts(url).pipe(take(options.limit))
    : matched.fetchProducts(url);

  obs.pipe( toArray() )
    .subscribe(products => {
      console.log(JSON.stringify({ products }));
    })
}