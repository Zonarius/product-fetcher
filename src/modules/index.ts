import { Fetcher } from '../model';
import * as decathlon from './decathlon';
import * as canyon from './canyon';

export default {
  decathlon, canyon
} as Record<string, Fetcher<any>>