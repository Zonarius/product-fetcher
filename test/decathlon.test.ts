import { getParsed } from "../src/fetch"
import { getProductListUrls, parseProduct } from "../src/modules/decathlon";

describe("decathlon", () => {
  test("parse single product", async () => {
    const doc = await getParsed("https://www.decathlon.at/sport-rennrad/301734-69509-rennrad-triban-rc520-disc.html");
    const parsed = parseProduct(doc);
    expect(parsed).toMatchSnapshot();
  })

  test("parse product list", async () => {
    const doc = await getParsed("https://www.decathlon.at/6873-rennrader");
    expect(getProductListUrls(doc).length).toBeGreaterThan(0);
  })
})