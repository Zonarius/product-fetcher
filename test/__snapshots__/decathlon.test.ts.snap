// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decathlon parse single product 1`] = `
Object {
  "name": "RENNRAD TRIBAN RC520 DISC",
  "price": 899.9,
  "technicalData": Object {
    "Antrieb/Schaltung": "Schalthebel Shimano 105 R7000, 
Umwerfer Shimano 105 R7000, 
Schaltwerk Shimano 105 R7000, 11-Gang mit langem Schaltkäfig, 
Gesamtgewicht: 843 g",
    "Bremsen": "Scheibenbremsen TRP HY/RD, mechanisch auf hydraulischen Kolben. Zuverlässig und effizient. 160 mm Scheibe vorne und hinten. 
Anti-Kompressionsbremszug JAGWIRE für präzises, wirksames Bremsen.
Halbmetallischer Bremsbelag TRP (Version M515/M525).",
    "DIE WAHL DER RICHTIGEN FAHRRADGRÖSSE": "Die Wahl der Fahrradgröße hängt nicht nur von deiner Körpergröße, sondern auch von deiner Morphologie ab (Arm- & Beinlänge). 
Wenn du zwischen zwei Größen bist, dann hast du bei Wahl des kleineren Modells eine sportlichere und bei Wahl des größeren eine komfortablere Position.
XS 154 165 cm
S 165 173 cm
M 173 180 cm
L 180 188 cm
XL 188 200 cm
Im Store kannst du ein Triban 500/520 testen, da die Rahmengeometrie identisch ist.
Einige Filialen bieten einen speziellen Vermessungsservice an.",
    "Eignet sich das RC520 zum Bikepacking?": "Ja, dieses Fahrrad wurde für Tourenfahrten auf der Straße wie beispielsweise das sogenannte „Bikepacking“ entwickelt. Das bedeutet, dass du beispielsweise am Sattel, Rahmen und Lenker leichte Taschen anbringen kannst. Dank seines hohen Komforts ist das RC520 dein idealer Begleiter während langer Tage im Sattel.
Dieses Fahrrad ist jedoch nicht auf längere Reisen ausgelegt. Von der Anbringung größerer Radtaschen an einem Gepäckträger raten wir ab.",
    "Finde heraus, wie das Fahrrad in der Praxis genutzt wird!": "Unsere Triban-Fahrräder haben bereits einige Abenteuer erlebt! Entdecke unsere Triban Storys auf https://triban.exposure.co/",
    "Gabel": "Neue Triban-Evo Gabel mit Holmen aus Carbon und Schaft aus Aluminium, 1 1/8 Zoll Aheadset. Diese Gabel verbindet Komfort mit geringem Gewicht (680 g) und Präzision.
Das Carbon wurde überarbeitet, um eine gute seitliche Steifigkeit und Filterung frontaler Vibration zu gewährleisten.
Aussparungen zum Anbringen eines Front-Gepäckträgers. Maximallast Gabel: 9 kg. 
Ösen an der Radachse zur Befestigung eines Schutzblechs. 
Kompatibel mit Reifen bis zu 700 × 36 oder 650 × 40.",
    "Gewicht": "10,4 kg in Größe M ohne Pedale",
    "Instructions": "Download",
    "Kompatible Schutzbleche": "Dieses Fahrrad ist mit den folgenden Schutzblechen aus unserem Sortiment kompatibel:
- SATTEL-SCHUTZBLECH RENNRAD FLASH BTWIN (Ref.8327786)
- SCHUTZBLECHSET RENNRAD RACEBLADE PRO XL VON SKS (Ref. 8515529)
- SCHUTZBLECH-SET RENNRAD BLUEMELS MATT 35 VON SKS (Ref. 8385659)

Gib die Referenznummer in die Suchfunktion deiner DECATHLON Seite ein, um nähere Informationen zu den Produkten zu erhalten.",
    "Laufräder": "Laufräder Triban Tubeless Ready* Light 
9 mm Achse mit Schnellspanner
Aluminium 6063T6
ETRTO Größe: 622 × 17C,
24 mm hoch für mehr seitliche Steifigkeit. Kreuzspeichen aus Stahl für mehr Steifigkeit (je 28 vorne und hinten).
Gedichtete Patronenlager.Gewicht pro Paar (ohne Kassette und ohne QR): 2.000 g.
*Zur Umrüstung auf Tubeless benötigst du ein Paar Tubeless Ready Reifen und ein Umrüstset mit 2 Tubeless Ventilen, 2 Felgenbändern und einer Flasche Pannenschutz-Flüssigkeit (Dichtmilch).",
    "Lenker/Vorbau/Steuersatz": "Ergonomisch optimierter Triban Alulenker mit ausgezeichneter Griffigkeit und hervorragendem Komfort.
Lenkerbreite (Mitte/Mitte): 
XS/S: 400 mm. M/L: 420 mm. XL: 440 mm.
Neue Triban Gabel:

XS/S: 80 mmM: 100 mm L: 110 mmXL: Länge: 120 mm.",
    "Pedale": "Inkl. FlachpedaleDas angegebene Gewicht des Fahrrads beinhaltet nicht die Pedale.",
    "Rahmen": "Neuer Triban-Evo Rahmen aus Aluminium 6061 T6, Komfortgeometrie mit angepasstem Sloping. 
Gewicht: 1780 g in Größe M. Stabiler, vielseitiger Rahmen: Nachrüsten von Schutzblechen und Gepäckträger möglich.
Kompatibel mit Reifen bis zu 700 × 36 oder 650 × 40.",
    "Reifen": "Reifen: Triban Resist+
410 g.
Skinwall Drahtreifen
mit Pannenschutz: 1 mm Nylon.
Robuste Karkasse: 55 TPI. Maße: 28–622 (700 × 28c) für noch höheren Komfort.
Max. Druck: 8 bar/116 psiDiese Reifen sind nicht Tubeless Ready.",
    "Sattel/Sattelstütze": "Neuer Sattel Triban ErgoFit. Hängematten-Konstruktion für mehr Komfort. Triban Sattelstütze aus Aluminium. Einfache Sattel-Einstellung, Durchmesser: 27,2 mm, Länge: 350 mm in M / L / XL, 300 mm in S, 250 mm in XS.",
    "Tretlager/Kassette": "MICROSHIFT Kassette CS-H110 11V 11/32 
(11/12/13/14/15/17/19/21/24/28/32).
Kompakter Shimano RS 510 50×34-Kurbelsatz. Länge der Kurbel nach Fahrradgröße: 
XS, S: 170 mm. M: 172,5 mm. L/XL: 175 mm.",
    "Zubehör/Ausstattung": "Lieferung inkl. Klingel",
  },
  "weight": 10.4,
}
`;
